src/bartiq/compilation/_evaluate.py

```diff
@@ -95,234 +49,42 @@ def evaluate(routine, assignments, *, backend=sympy_backend, functions_map=None)
-    backend: SymbolicBackend[T_expr],
-    functions_map: Optional[FunctionsMap] = None,
-) -> Routine:
-    for assignment in assignments:
-        routine = _evaluate_routine_over_assignment(routine, assignment, backend, functions_map)
-    return routine
-
-
-def _evaluate_routine_over_assignment(
-    routine: Routine,
-    assignment: Assignment,
-    backend: SymbolicBackend[T_expr],
-    functions_map: Optional[FunctionsMap] = None,
-) -> Routine:
-    """Dispatches operation assignment based upon the assignment type."""
-    # First, check that the assignment is to a number
-    if not isinstance(assignment.value, NUMBER_TYPES):
-        raise BartiqCompilationError(
-            f"Can only evaluate variables to numbers; attempted assignment {assignment.variable} = {assignment.value}"
-        )
-    # Evaluate assignment based on type
-    if isinstance(assignment, _VariableAssignment):
-        routine = _evaluate_routine_over_assignment_input_variable(routine, assignment, backend, functions_map)
-    if isinstance(assignment, _RegisterSizeAssignment):
-        assert assignment.direction == PortDirection.input
-        routine = _evaluate_routine_over_assignment_input_register_size(routine, assignment, backend, functions_map)
-
-    return routine
-
-
-def _evaluate_routine_over_assignment_input_variable(
-    routine: Routine,
-    assignment: _VariableAssignment,
-    backend: SymbolicBackend[T_expr],
-    functions_map: Optional[FunctionsMap] = None,
-) -> Routine:
-    """Evaluates a routine over a single input variable assignment."""
-    variable = assignment.variable
-    value = assignment.value
-
-    # Deal with case when routine doesn't reference input variable
-    if variable not in routine.input_params:
-        return routine
-
-    # Parse the routine and the assignment
-    old_function = to_symbolic_function(routine, backend)
-
-    # Evaluate function over assignment and substitute any user-provided functions
-    new_function = define_expression_functions(old_function, functions_map, strict=False)
-    new_function = evaluate_function_at(new_function, variable, value, backend)
-
-    update_routine_with_symbolic_function(routine, new_function)
-    return routine
-
-
-def _evaluate_routine_over_assignment_input_register_size(
-    routine: Routine,
-    assignment: _RegisterSizeAssignment,
-    backend: SymbolicBackend[T_expr],
-    functions_map: Optional[FunctionsMap] = None,
-) -> Routine:
-    """Evaluates a routine over a single input register size assignment."""
-    register = f"#{assignment.register}"
-    value = assignment.value
-
-    old_function = to_symbolic_function(routine, backend)
-
-    new_function = define_expression_functions(old_function, functions_map)
-    new_function = set_input_port_size_to_constant_value(new_function, register, value, backend)
-
-    update_routine_with_symbolic_function(routine, new_function)
-    return routine
+def _evaluate_internal(
```

Comment:
(see my previous comment first)
The `evaluate` is a this wrapper around `_evaluate_inner`, which parses constants in assignments and validates inputs. `evaluate_inner` is the actual implementation. The distinction is important, because `_evaluate_inner` is called recursively, and we don't want to perform input validation with every call.